### ISSUE TITLE
lxc-attach: Permission denied for container with AppArmor profile and no_new_privs flag set

### ISSUE THREAD

**Issue by alivenets** [lxc/lxc#3393]
# Required information

 * Distribution: Ubuntu
 * Distribution version: 18.04
 * The output of
   * `lxc-start --version`: 3.1.0
   * `lxc-checkconfig`
```
Kernel configuration not found at /proc/config.gz; searching...
Kernel configuration found at /boot/config-4.15.0-30-generic
--- Namespaces ---
Namespaces: enabled
Utsname namespace: enabled
Ipc namespace: enabled
Pid namespace: enabled
User namespace: enabled
Network namespace: enabled

--- Control groups ---
Cgroups: enabled

Cgroup v1 mount points: 
/sys/fs/cgroup/systemd
/sys/fs/cgroup/net_cls,net_prio
/sys/fs/cgroup/blkio
/sys/fs/cgroup/devices
/sys/fs/cgroup/perf_event
/sys/fs/cgroup/cpuset
/sys/fs/cgroup/memory
/sys/fs/cgroup/rdma
/sys/fs/cgroup/cpu,cpuacct
/sys/fs/cgroup/pids
/sys/fs/cgroup/freezer
/sys/fs/cgroup/hugetlb

Cgroup v2 mount points: 
/sys/fs/cgroup/unified

Cgroup v1 clone_children flag: enabled
Cgroup device: enabled
Cgroup sched: enabled
Cgroup cpu account: enabled
Cgroup memory controller: enabled
Cgroup cpuset: enabled

--- Misc ---
Veth pair device: enabled, not loaded
Macvlan: enabled, not loaded
Vlan: enabled, not loaded
Bridges: enabled, not loaded
Advanced netfilter: enabled, not loaded
CONFIG_NF_NAT_IPV4: enabled, not loaded
CONFIG_NF_NAT_IPV6: enabled, not loaded
CONFIG_IP_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_IP6_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_NETFILTER_XT_TARGET_CHECKSUM: enabled, not loaded
CONFIG_NETFILTER_XT_MATCH_COMMENT: enabled, not loaded
FUSE (for use with lxcfs): enabled, not loaded

--- Checkpoint/Restore ---
checkpoint restore: enabled
CONFIG_FHANDLE: enabled
CONFIG_EVENTFD: enabled
CONFIG_EPOLL: enabled
CONFIG_UNIX_DIAG: enabled
CONFIG_INET_DIAG: enabled
CONFIG_PACKET_DIAG: enabled
CONFIG_NETLINK_DIAG: enabled
File capabilities: 

Note : Before booting a new kernel, you can check its configuration
usage : CONFIG=/path/to/config /usr/bin/lxc-checkconfig
```
   * `uname -a`: `Linux ubuntu-bionic 4.15.0-30-generic #32-Ubuntu SMP Thu Jul 26 17:42:43 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`

   * `cat /proc/self/cgroup`
```
12:hugetlb:/
11:freezer:/
10:pids:/user.slice/user-1000.slice/session-7.scope
9:cpu,cpuacct:/user.slice
8:rdma:/
7:memory:/user.slice
6:cpuset:/
5:perf_event:/
4:devices:/user.slice
3:blkio:/user.slice
2:net_cls,net_prio:/
1:name=systemd:/user.slice/user-1000.slice/session-7.scope
0::/user.slice/user-1000.slice/session-7.scope
```
   * `cat /proc/1/mounts`
```
sysfs /sys sysfs rw,nosuid,nodev,noexec,relatime 0 0
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
udev /dev devtmpfs rw,nosuid,relatime,size=491828k,nr_inodes=122957,mode=755 0 0
devpts /dev/pts devpts rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 0 0
tmpfs /run tmpfs rw,nosuid,noexec,relatime,size=100896k,mode=755 0 0
/dev/sda1 / ext4 rw,relatime,data=ordered 0 0
securityfs /sys/kernel/security securityfs rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /dev/shm tmpfs rw,nosuid,nodev 0 0
tmpfs /run/lock tmpfs rw,nosuid,nodev,noexec,relatime,size=5120k 0 0
tmpfs /sys/fs/cgroup tmpfs ro,nosuid,nodev,noexec,mode=755 0 0
cgroup /sys/fs/cgroup/unified cgroup2 rw,nosuid,nodev,noexec,relatime 0 0
cgroup /sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,name=systemd 0 0
pstore /sys/fs/pstore pstore rw,nosuid,nodev,noexec,relatime 0 0
cgroup /sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_cls,net_prio 0 0
cgroup /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
cgroup /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
cgroup /sys/fs/cgroup/perf_event cgroup rw,nosuid,nodev,noexec,relatime,perf_event 0 0
cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset 0 0
cgroup /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory 0 0
cgroup /sys/fs/cgroup/rdma cgroup rw,nosuid,nodev,noexec,relatime,rdma 0 0
cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
cgroup /sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
cgroup /sys/fs/cgroup/hugetlb cgroup rw,nosuid,nodev,noexec,relatime,hugetlb 0 0
systemd-1 /proc/sys/fs/binfmt_misc autofs rw,relatime,fd=36,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=11655 0 0
hugetlbfs /dev/hugepages hugetlbfs rw,relatime,pagesize=2M 0 0
debugfs /sys/kernel/debug debugfs rw,relatime 0 0
mqueue /dev/mqueue mqueue rw,relatime 0 0
configfs /sys/kernel/config configfs rw,relatime 0 0
fusectl /sys/fs/fuse/connections fusectl rw,relatime 0 0
lxcfs /var/lib/lxcfs fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
vagrant /vagrant vboxsf rw,nodev,relatime 0 0
tmpfs /run/user/1000 tmpfs rw,nosuid,nodev,relatime,size=100892k,mode=700,uid=1000,gid=1000 0 0
```

# Issue description

Attach to container with config `lxc.apparmor.profile = lxc-contaner-default-cgns` and `lxc.no_new_privs = 1` from `lxc-attach` confined by AppArmor profile fails with `lxc-attach: TestBusybox: lsm/lsm.c: lsm_process_label_set_at: 174 Operation not permitted - Failed to set AppArmor label "lxc-container-default-cgns"
`

# Steps to reproduce

 1. Create privileged container with `lxc-create -t busybox TestBusybox`
 2. Add following configuration to `/var/lib/lxc/TestBusybox/config`:
```
lxc.apparmor.profile = lxc-container-default-cgns
lxc.no_new_privs = 1
```
 3. Install AppArmor profile for `lxc-attach` (in the bottom):
```
sudo cp usr.bin.lxc-attach.txt /etc/apparmor.d/usr.bin.lxc-attach
sudo apparmor_parser -r -W /etc/apparmor.d/usr.bin.lxc-attach
```
 4. Start container: `lxc-start TestBusybox`
 5. Attach to container: `lxc-attach TestBusybox`
 6. The failure happens

# Information to attach

 - [x] any relevant kernel output (`dmesg`)
```
[ 2031.882698] audit: type=1400 audit(1587677011.912:94): apparmor="DENIED" operation="change_profile" info="no new privs" error=-1 profile="/usr/bin/lxc-attach" name="lxc-container-default-cgns" pid=16484 comm="lxc-attach" target="lxc-container-default-cgns"
```
 - [x] container log (The <log> file from running `lxc-start -n <c> -l TRACE -o <logfile> `)
[attach.log](https://github.com/lxc/lxc/files/4525390/attach.log)
 - [x] the containers configuration file
```
# Template used to create this container: /usr/share/lxc/templates/lxc-busybox
# Parameters passed to the template:
# For additional config options, please look at lxc.container.conf(5)

# Uncomment the following line to support nesting containers:
#lxc.include = /usr/share/lxc/config/nesting.conf
# (Be aware this has security implications)

#lxc.net.0.type = veth
#lxc.net.0.link = lxcbr0
#lxc.net.0.flags = up
#lxc.net.0.hwaddr = 00:16:3e:8f:e1:5f
lxc.rootfs.path = dir:/usr/var/lib/lxc/TestBusybox/rootfs
lxc.signal.halt = SIGUSR1
lxc.signal.reboot = SIGTERM
lxc.uts.name = "TestBusybox"
lxc.tty.max = 1
lxc.pty.max = 1
# Disabled capabilities dropping
# lxc.cap.drop = sys_module mac_admin mac_override sys_time

# When using LXC with apparmor, uncomment the next line to run unconfined:
lxc.apparmor.profile = lxc-container-default-cgns
lxc.no_new_privs = 1 

lxc.mount.auto = cgroup:mixed proc:mixed sys:mixed
lxc.mount.entry = shm /dev/shm tmpfs defaults 0 0
lxc.mount.entry = /lib lib none ro,bind 0 0
lxc.mount.entry = /usr/lib usr/lib none ro,bind 0 0
lxc.mount.entry = /lib64 lib64 none ro,bind 0 0
lxc.mount.entry = /sys/kernel/security sys/kernel/security none ro,bind,optional 0 0
```
- [x] lxc-attach AppArmor profile
[usr.bin.lxc-attach.txt](https://github.com/lxc/lxc/files/4525409/usr.bin.lxc-attach.txt)

**Comment by brauner** [comment]
Works on 4.* just fine and 3.1 is out of support. :/
If you can reproduce this on one of our LTS releases, i.e. 3.0.4 or 4.0.1?
```
lxc-start b4 20200503121622.737 NOTICE   conf - conf.c:lxc_setup:3433 - The container "b4" is set up
lxc-start b4 20200503121622.737 INFO     lsm - lsm/lsm.c:lsm_process_label_set_at:157 - Set AppArmor label to "lxc-container-default-cgns"
lxc-start b4 20200503121622.737 INFO     apparmor - lsm/apparmor.c:apparmor_process_label_set:1185 - Changed AppArmor profile to lxc-container-default-cgns
lxc-start b4 20200503121622.737 DEBUG    start - start.c:do_start:1250 - Set PR_SET_NO_NEW_PRIVS to block execve() gainable privileges
```
and then attach:
```
lxc-attach b4 20200503121628.666 TRACE    attach - attach.c:attach_child_main:785 - Set PR_SET_NO_NEW_PRIVS
lxc-attach b4 20200503121628.666 INFO     lsm - lsm/lsm.c:lsm_process_label_set_at:157 - Set AppArmor label to "lxc-container-default-cgns"
lxc-attach b4 20200503121628.666 TRACE    attach - attach.c:attach_child_main:799 - Set AppArmor LSM label to "lxc-container-default-cgns"
```

**Comment by brauner** [comment]
I'll let @stgraber take another look at this but this rather looks like you're messing with AppArmor directly which shouldn't be necessary on stock Ubuntu 18.04 at all. So it's likely that you're just misconfiguring something.

**Comment by alivenets** [comment]
Thanks for your attention to the issue. Reproduced for me with LXC 4.0.1. Regarding distro, It is stock Ubuntu 18.04 + AppArmor profile for `lxc-attach`. Basically, my case is to attach to container using C API from the program confined by AppArmor profile. I reproduced the issue with built-in LXC tools with additional AppArmor profile.

Several questions:
1. Why is `lxc-attach` not confined by AppArmor profile?
2. In `start.c` (https://github.com/lxc/lxc/blob/master/src/lxc/start.c#L1243) no_new_privs flag is set after AppArmor profile set, but in `attach.c` (https://github.com/lxc/lxc/blob/master/src/lxc/attach.c#L777) the order is inverted. Is it done on intention?
